### PR TITLE
Mixed grid updates

### PR DESCRIPTION
**Pull request by JimBrouzoulis**
Some changes in the code to allow for non-contiguous ordering of cell types.

### CODE REVIEW

File: src/Dofs/MixedDofHandler.jl [hunk 1]
```diff
@@ -4,30 +4,63 @@ mutable struct FieldHandler
...
 
+
+struct CellDofs
+    dofs::Vector{Int}
+    offset::Vector{Int}
+    length::Vector{Int}
+end
+
+import Base.getindex
+function getindex(eldofs::CellDofs, el::Int64)
```
Comment (KristofferC) [inline src/Dofs/MixedDofHandler.jl]:
```suggestion
function Base.getindex(eldofs::CellDofs, el::Int64)
```

File: src/Dofs/MixedDofHandler.jl [hunk 2]
```diff
@@ -4,30 +4,63 @@ mutable struct FieldHandler
...
 end
 
+
+struct CellDofs
+    dofs::Vector{Int}
+    offset::Vector{Int}
+    length::Vector{Int}
+end
+
+import Base.getindex
```
Comment (KristofferC) [inline src/Dofs/MixedDofHandler.jl]:
```suggestion
```

File: src/Dofs/MixedDofHandler.jl [hunk 3]
```diff
@@ -4,30 +4,63 @@ mutable struct FieldHandler
...
+function getindex(eldofs::CellDofs, el::Int64)
+     return eldofs.dofs[eldofs.offset[el]:eldofs.offset[el]+eldofs.length[el]-1]
```
Comment (KristofferC) [inline src/Dofs/MixedDofHandler.jl]: Hoist `eldofs.offset[el]` into a variable to not lookup twice?

File: src/Dofs/MixedDofHandler.jl [hunk 4]
```diff
@@ -4,30 +4,63 @@ mutable struct FieldHandler
...
 
 function MixedDofHandler(grid::MixedGrid)
-    MixedDofHandler(FieldHandler[], Int[], Int[], JuAFEM.ScalarWrapper(false), grid)
+    MixedDofHandler(FieldHandler[], CellDofs([],[],[]), JuAFEM.ScalarWrapper(false), grid)
 end
 
 getfieldnames(fh::FieldHandler) = [field.name for field in fh.fields]
 getfielddims(fh::FieldHandler) = [field.dim for field in fh.fields]
 getfieldinterpolations(fh::FieldHandler) = [field.interpolation for field in fh.fields]
+import JuAFEM.ndofs_per_cell, JuAFEM.celldofs, JuAFEM.isclosed, JuAFEM.ndofs
```
Comment (KristofferC) [inline src/Dofs/MixedDofHandler.jl]: No need to import these, right?

File: src/Dofs/MixedDofHandler.jl [hunk 5]
```diff
@@ -4,30 +4,63 @@ mutable struct FieldHandler
...
+ndofs_per_cell(dh::MixedDofHandler, cell::Int64) = dh.cell_dofs.length[cell]
+ndofs(dh::MixedDofHandler) = maximum(dh.cell_dofs.dofs)
```
Comment (KristofferC) [inline src/Dofs/MixedDofHandler.jl]:
```suggestion
ndofs(dh::MixedDofHandler) = length(dh.cell_dofs.dofs)
```

If we are assuming that the celldofs start from 1 and are continous?

File: src/Dofs/MixedDofHandler.jl [hunk 6]
```diff
@@ -139,9 +192,14 @@ function _close!(dh::MixedDofHandler{dim}, cellnumbers, field_names, field_dims,
 
         end
         # after done creating dofs for the cell, push them to the global list
-        push!(dh.cell_dofs, cell_dofs...)
+        push!(dh.cell_dofs.dofs, cell_dofs...)
+        dh.cell_dofs.length[ci] = length(cell_dofs)
+        #push!(dh.cell_dofs, cell_dofs...)
         # push! the first index of the next cell to the offset vector
-        push!(dh.cell_dofs_offset, length(dh.cell_dofs)+1)
+        #dh.cell_dofs_offset[ci] = nextdof
```
Comment (KristofferC) [inline src/Dofs/MixedDofHandler.jl]: Should commented lines be deleted?

File: src/Dofs/MixedDofHandler.jl (hunk 4 above)
Comment (JimBrouzoulis) [reply src/Dofs/MixedDofHandler.jl]: Yup, no need for them.

File: src/Dofs/MixedDofHandler.jl (hunk 5 above)
Comment (JimBrouzoulis) [reply src/Dofs/MixedDofHandler.jl]: They are not continuous, simply an array with all the dofs for each cell after one another.
But I can't see `ndofs` being called frequently. Only in some initializations like for the sparse structure.

File: src/Dofs/MixedDofHandler.jl (hunk 6 above)
Comment (JimBrouzoulis) [reply src/Dofs/MixedDofHandler.jl]: Yes

File: src/Dofs/MixedDofHandler.jl (hunk 5 above)
Comment (KristofferC) [reply src/Dofs/MixedDofHandler.jl]: Yeah, I mixed it up with `ndofs_per_cell` -.-.

Still, I wonder if it wouldn't be better to cache it on `close!` in `dh.ndofs` or something.

Comment (JimBrouzoulis) [reply src/Dofs/MixedDofHandler.jl]: Sure, we can do that. Saving an integer may be acceptable use of memory in this case 😄

**Pull request merged**